src/main/java/app/sami/languageWeb/contract/ContractService.java

```diff
@@ -0,0 +1,90 @@
+package app.sami.languageWeb.contract;
+
+import app.sami.languageWeb.contract.dtos.StorageUriDto;
+import app.sami.languageWeb.contract.models.Contract;
+import app.sami.languageWeb.contract.models.Status;
+import app.sami.languageWeb.error.exceptions.ContractTranslatedFileAbsent;
+import app.sami.languageWeb.error.exceptions.NotFoundException;
+import app.sami.languageWeb.error.exceptions.UserNotAllowedException;
+import app.sami.languageWeb.request.RequestRepository;
+import app.sami.languageWeb.request.models.Request;
+import app.sami.languageWeb.storage.Storage;
+import lombok.AllArgsConstructor;
+import lombok.NoArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+import java.util.List;
+import java.util.UUID;
+
+@Service
+@AllArgsConstructor
+public class ContractService {
+
+    private final RequestRepository requestRepository;
+    private final Storage storage;
+    private final ContractRepository contractRepository;
+
+    public StorageUriDto getUploadUri(UUID subject, String fileName){
+        String path = String.format("/contracts/%s/%s__%s", subject, UUID.randomUUID(), fileName);
+        String url = storage.getUploadPresignedUrl(path);
+
+        return StorageUriDto.builder()
+                .url(url)
+                .fileName(path)
+                .build();
+    }
+
+    public Contract createContract(UUID subject, Long requestId){
+        Request request = requestRepository.findById(requestId).orElseThrow(NotFoundException::new);
+        if (request.isUser(subject)){
+            throw new UserNotAllowedException();
+        }
+        requestRepository.save(request.withStatus(app.sami.languageWeb.request.models.Status.ACCEPTED));
+        return contractRepository.save(Contract.builder()
+                .contractedUserId(subject)
+                .status(Status.PENDING)
+                .requestId(requestId)
+                .build());
+    }
+
+    public Contract updateContractParticipantStatus(UUID subject, Long contractId, Status status){
+        Contract contract = contractRepository.findById(contractId).orElseThrow(NotFoundException::new);
+        if (status.equals(Status.COMPLETED) && !contract.isFileAdded()){
+            throw new ContractTranslatedFileAbsent();
+        }
+
+        contract = contractRepository.save(contract.updateStatus(subject, status));
+
+        if (contract.isContractedStatus(Status.CANCELLED) && contract.isContractorStatus(Status.CANCELLED)){
+            return cancelContract(contractId);
```

Comment:
you are making 2 saves, why ? That's something the `contract.updateStatus` can do:

```java
public Contract updateStatus(UUID subject, ContractStatus status) {
  if (whatever) {
    throw new notallowed();
  }  
  if (status == cancelled) {
    // cancel for both
    return this;
  }
  
  // set for the right person
  return this;
}
```

Reply:
if you'd rather keep the object immutable you can do `return this.withStatus(cancelled).withOtherStatus(cancelled)`